### ISSUE TITLE
DNS resolution periodically fails in containers

### ISSUE THREAD

**Issue by mattdodge** [rancher/rancher#9894]
**Rancher versions:**
rancher/server: 1.6.9
rancher/agent: 1.2.6

**Infrastructure Stack versions:**
healthcheck: 0.3.3
ipsec: 0.1.4
network-services: 0.2.6
scheduler: 0.6.3
rancher/dns: 0.15.3

**Docker version: (`docker version`,`docker info` preferred)**
```
Client:
 Version:      17.03.0-ce
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   3a232c8
 Built:        Tue Feb 28 08:01:32 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.03.0-ce
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   3a232c8
 Built:        Tue Feb 28 08:01:32 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Operating system and kernel: (`cat /etc/os-release`, `uname -r` preferred)**
```
NAME="Ubuntu"
VERSION="16.04.1 LTS (Xenial Xerus)"
4.4.0-93-generic
```

**Type/provider of hosts: (VirtualBox/Bare-metal/AWS/GCE/DO)**
AWS EC2

**Setup details: (single node rancher vs. HA rancher, internal DB vs. external DB)**
HA Rancher, External DB

**Environment Template: (Cattle/Kubernetes/Swarm/Mesos)**
Cattle

**Steps to Reproduce:**
Tail the `rancher/dns` container logs. Once a bunch of "No answer found" messages start appearing, in another container on the host try to resolve DNS for anything (e.g., `ping google.com`).

**Results:**
Watching the dns container logs, eventually an error like this appears. This seems to kick of the error state.
```
Failed to generate answers: Get http://localhost/2015-12-19/services: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
```
Then, we see a slew of "No answer found" messages like so come through:
```
No answer found" client=10.42.171.6 question=5-17-1-app.agent.datadoghq.com.rancher.internal. type=AAAA
```
We'll see a bunch of different domains being questioned too, not just the datadog one in the example. It's not clear, to me at least, whether these are errors in lookups happening **during** the error state or if these are **causing** the error state, but when these are happening any DNS lookups in containers are failing.

Eventually, the errors stop and we see a message like this, and DNS resolution is working again.
```
time="2017-09-16T16:31:31Z" level=info msg="Reloading answers"
time="2017-09-16T16:31:31Z" level=info msg="Reloaded answers"
```

However, a few minutes later it all starts over again. So, a summary of the logs we see looks like this:
```
time="2017-09-16T16:31:53Z" level=error msg="Failed to generate answers: Get http://localhost/2015-12-19/services: net/http: request canceled (Client.Timeout exceeded while awaiting headers)"
time="2017-09-16T16:31:53Z" level=info msg="Reloading answers"
time="2017-09-16T16:31:53Z" level=info msg="Reloaded answers"
time="2017-09-16T16:31:53Z" level=info msg="No answer found" client=10.42.113.12 question=dns.that.I.want.to.lookup. type=A
time="2017-09-16T16:31:53Z" level=info msg="No answer found" client=10.42.113.12 question=dns.that.I.want.to.lookup. type=A
time="2017-09-16T16:31:53Z" level=info msg="No answer found" client=10.42.113.12 question=dns.that.I.want.to.lookup.rancher.internal. type=A
..... repeats many times .....
time="2017-09-16T16:34:37Z" level=info msg="Reloading answers"
time="2017-09-16T16:34:37Z" level=info msg="Reloaded answers"
```

**Comment by mattdodge** [comment]
Update, maybe helpful. When the "error state" occurs the contents of `/etc/rancher-dns/answers.json` are just the characters `null`. It's only 4 bytes of a file. That explains the "No answer found" that happens.

It really seems like the problem is that it can't generate the answers and that timeout error. Any ideas what might cause this?
```
Get http://localhost/2015-12-19/services: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
```

**Comment by aemneina** [comment]
@mattdodge do these dns outages occur when service updates happen? I believe there could be a blip in service as the answers get reloaded/processed.

**Comment by mattdodge** [comment]
@aemneina It may have been caused by a service update but the errors continued to happen well beyond the time of the update. I will say we've been closely monitoring this since the issue was reported and we haven't seen it occur since then. There was a solid 24 hours though when we were getting these errors almost 100% reliably repeatable in one of our environments.

**Comment by ptempier** [comment]
@mattdodge 
Hi, did you managed to find the cause of this problem? Do you srtill experiment this issue?

**Comment by mattdodge** [comment]
Have not experienced the issue again but to be honest we haven't been running this Rancher cluster anymore for a few months. Never ended up finding the cause of the issue though

**Comment by sm6yvr** [comment]
Hi, I have the exact same issue with rancher 1.6.10. Does anyone have a proper fix or a workaround?

**Comment by ptempier** [comment]
Hi you d need to compile by yourself :
https://github.com/rancher/rancher-dns/pull/71
This basically mitigate the metadata issue by 90%

Another fix that's quite efficient it to put a sleep loop at your container startup so that it doesn't do anything until it can see itself in the metadata.
You can put this loop before anything, as a lot of thing wont work of the metadata can t see the container it s supposed to manage.

**Comment by eesprit** [comment]
@ptempier 
> Another thing that's quite efficient it to put a sleep loop at your container startup so that it doesn't do anything until it can see itself in the metadata.
> You can put this loop before anything, as a lot of thing wont work of the metadata can t see the container it s supposed to manage.

Shouldn't it be done in the **/.r/r** "meta-entrypoint" then ?
By the way I couldn't find which github repos is hosting the source code of that /.r/r binary, I am indeed curious to see what it does.

**Comment by leodotcloud** [comment]
@mattdodge You are right. The DNS timeouts are fixed by https://github.com/rancher/rancher-dns/pull/71. The fix is available in v1.6.15, please upgrade and keep us posted if you are still experiencing the problem.

The timeout (while connecting to metadata) issue could be due to CPU load on the machine.

**Comment by deniseschannon** [comment]
With the release of Rancher 2.0, development on v1.6 is only limited to critical bug fixes and security patches.
